### ISSUE TITLE
Can we convert collate_fn in Dataloader to a custom class(or function) ?

### ISSUE THREAD

**Issue by wlhgtc** [allenai/allennlp#5218]
For some task(such as masking ), we want to mask different pos in different epoch for the same record.
 Like the [code](https://github.com/huggingface/transformers/blob/1c06240e1b3477728129bb58e7b6c7734bb5074e/src/transformers/trainer.py#L546) in Transformers, we could pass a class to the collate_fn. 
But now we only have a simple [allennlp_collate](https://github.com/allenai/allennlp/blob/b180d5759f58aa98108b8725d27e926887d67856/allennlp/data/data_loaders/data_loader.py#L17).
The old version of our Dataloader  is a subclass of Pytorch Dataloder, seems we write our own one in the latest version ?

**Comment by wlhgtc** [comment]
@epwalsh

**Comment by wlhgtc** [comment]
Maybe just like what we have done in [optimizer](https://github.com/allenai/allennlp/blob/main/allennlp/training/optimizers.py), import `Collator` class from Transformers and use a custom base class to cover ?

**Comment by epwalsh** [comment]
This would be a good feature to have, and easy to add without breaking backwards compat.

Would you like to make a PR?

**Comment by wlhgtc** [comment]
> This would be a good feature to have, and easy to add without breaking backwards compat.
> 
> Would you like to make a PR?

Well, I'll try it soon.

**Comment by dirkgr** [comment]
Done! Thank you!
